src/textual/screen.py

```diff
@@ -293,18 +294,52 @@ def focus_chain(self) -> list[Widget]:
 
         widgets: list[Widget] = []
         add_widget = widgets.append
-        stack: list[Iterator[Widget]] = [iter(self.focusable_children)]
-        pop = stack.pop
-        push = stack.append
+        focus_sorter = attrgetter("_focus_sort_key")
+        node_stack: list[Iterator[Widget]] = [
+            iter(sorted(self.displayed_children, key=focus_sorter))
+        ]
+        pop_node = node_stack.pop
+        push_node = node_stack.append
+
+        # Instead of relying on the property `DOMNode.visible`, we manually keep track
+        # of the visibility of the DOM to save on DOM traversals inside `.visible`.
+        visibility_stack: list[bool] = [self.visible]
+        pop_visibility = visibility_stack.pop
+        push_visibility = visibility_stack.append
 
-        while stack:
-            node = next(stack[-1], None)
+        while node_stack:
+            node = next(node_stack[-1], None)
             if node is None:
-                pop()
+                pop_node()
+                pop_visibility()
             else:
+                node_visibility = node.styles.get_rule("visibility")
                 if node.is_container and node.can_focus_children:
-                    push(iter(node.focusable_children))
-                if node.focusable:
+                    sorted_displayed_children = sorted(
+                        node.displayed_children, key=focus_sorter
+                    )
+                    push_node(iter(sorted_displayed_children))
+                    # When we push a new container to the stack, we need to update
+                    # the visibility stack.
+                    if node_visibility is None:
+                        # If the node doesn't have explicit visibility set, we inherit
+                        # it from the stack, unless we're at the top of the stack,
+                        # in which case we must compute the node visibility.
+                        push_visibility(
+                            visibility_stack[-1] if visibility_stack else node.visible
+                        )
+                    else:
+                        push_visibility(node_visibility != "hidden")
+                node_is_visible = (
+                    node_visibility != "hidden"
+                    if node_visibility
+                    else visibility_stack[-1]
+                )
+                if (  # Same as `if node.focusable`, but we cache inherited visibility.
+                    node_is_visible
+                    and node.can_focus
+                    and not node._self_or_ancestors_disabled
```

Comment:
The  `_self_or_ancestors_disabled` method will look up the DOM to the root, which makes it a little inefficient to call when we are recursing the entire DOM.

If a branch of the tree is disabled, then all its children will be disabled. So we could avoid traversing the remaining tree if we encounter a disabled widget.

Reply:
🤦  of course!

Implemented that in ae773ed.